test/box-luatest/gh_8505_synchro_triggers_test.lua

```diff
@@ -0,0 +1,65 @@
+local t = require('luatest')
+local replica_set = require('luatest.replica_set')
+local server = require('luatest.server')
+
+local g = t.group('gh-8505-synchro-triggers')
+
+g.before_all(function(g)
+    g.replica_set = replica_set:new({})
+    g.box_cfg = {
+        replication_timeout = 0.01,
+        replication = {
+            server.build_listen_uri('server1', g.replica_set.id),
+            server.build_listen_uri('server2', g.replica_set.id),
+        },
+    }
+
+    g.box_cfg.election_mode = 'voter'
+    g.server2 = g.replica_set:build_and_add_server({
+        alias = 'server2', box_cfg = g.box_cfg
+    })
+
+    g.box_cfg.election_mode = 'candidate'
+    g.server1 = g.replica_set:build_and_add_server({
+        alias = 'server1', box_cfg = g.box_cfg
+    })
+
+    g.replica_set:start()
+    g.server1:wait_for_election_leader()
+    g.server1:exec(function()
+        box.schema.create_space('test', {is_sync = true}):create_index('pk')
+    end)
+    g.server2:wait_for_vclock_of(g.server1)
+end)
+
+g.after_all(function(g)
+    g.replica_set:drop()
+end)
+
+g.test_on_commit_trigger = function(g)
+    g.server1:exec(function()
+        box.begin()
+        box.on_commit(function(iter) iter() end)
+        box.space.test:upsert({1}, {{'=', 1, 1}})
+        box.commit()
+    end)
+end
+
+g.test_on_rollback_trigger = function(g)
+    -- Force ACK gathering to fail and cause rollback
+    g.box_cfg.replication_synchro_timeout = 1e-9
```

Comment:
When I run the test like this `python test-run.py $(yes 8505 | head -n 30)`, I get some failures:
```
[006] Test failed! Output from reject file /tmp/t/rejects/box-luatest/gh_8505_synchro_triggers.reject:
[006] Tarantool version is 3.0.0-entrypoint-290-g1bb7177b1
[006] TAP version 13
[006] 1..2
[006] # Started on Wed Jun  7 22:35:04 2023
[006] # Starting group: gh-8505-synchro-triggers
[006] ok     1	gh-8505-synchro-triggers.test_on_commit_trigger
[006] not ok 2	gh-8505-synchro-triggers.test_on_rollback_trigger
[006] #   ...ntool/test/box-luatest/gh_8505_synchro_triggers_test.lua:59: attempt to index local 'err' (a nil value)
[006] #   stack traceback:
[006] #   	...ntool/test/box-luatest/gh_8505_synchro_triggers_test.lua:54: in function 'gh-8505-synchro-triggers.test_on_rollback_trigger'
[006] #   	...
[006] #   	[C]: in function 'xpcall'
[006] # Ran 2 tests in 2.468 seconds, 1 succeeded, 1 errored
```

Reply:
For some reason, transactions with tiny `replication_synchro_timeout` sometimes manage to complete. This doesn't happen, when connection is explicitly broken with proxy.  I'll investigate this behavior, for this patch, I suppose, we can stay with proxies

Reply:
Still fails:
```
[015] replication-luatest/gh_8505_synchro_triggers_t>                 [ fail ]
[015] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/gh_8505_synchro_triggers.reject:
[015] Tarantool version is 3.0.0-entrypoint-306-gde0731bd3
[015] TAP version 13
[015] 1..3
[015] # Started on Wed Jun 14 22:20:55 2023
[015] # Starting group: gh-8505-synchro-triggers
[015] ok     1	gh-8505-synchro-triggers.test_on_commit_trigger
[015] ok     2	gh-8505-synchro-triggers.test_on_rollback_trigger
[015] not ok 3	gh-8505-synchro-triggers.test_session_on_commit
[015] #   Quorum collection for a synchronous transaction is timed out
[015] #   stack traceback:
[015] #   	...st/replication-luatest/gh_8505_synchro_triggers_test.lua:82: in function 'gh-8505-synchro-triggers.test_session_on_commit'
[015] #   	...
[015] #   	[C]: in function 'xpcall'
[015] # Ran 3 tests in 2.496 seconds, 2 succeeded, 1 errored
```

Reply:
Managed to repoduce it. Should be stable from now on